# Checkpoint 5 : Final Evaluation on the Test Set


## 1. All works

In [74]:
import pandas as pd
import numpy as np
import re
from sklearn.model_selection import train_test_split
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.impute import SimpleImputer
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

pd.set_option("display.max_columns", 70)


In [75]:
df = pd.read_csv("house_sale.csv")

In [76]:
df

,id_x,rel_url,estate_rel_url_x,datetime_scrape_x,price,currency_x,location,attributes,city_when,city,day_x,hour_x,repair,vip,featured,products_label,bill_of_sale,mortgage,img_url,id_y,estate_id,estate_rel_url_y,datetime_scrape_y,description,unit_price,total_price,currency_y,owner_name,owner_title,shop_name,shop_title,address,lat,lng,updated,views,day_y,hour_y,estate_details_id_x,Binanın növü,Kateqoriya,Mərtəbə,Otaq sayı,Sahə,Torpaq sahəsi,Təmir,Çıxarış,İpoteka,estate_details_id_y,estate_rel_url,extra_info
0,5df36281-6dc6-4d5d-89a7-5fcfa86f7608,/alqi-satqi?page=174,/items/4521724,2024-10-05 22:07:37.60613+00,499999.0,AZN,Səbail r.,"4 otaqlı, 145 m², 7/9 mərtəbə","Bakı, dünən 23:52",bakı,05.10.2024,23:52,Təmirli,vipped,featured,NaN,Çıxarış var,NaN,https://bina.azstatic.com/uploads/f460x345/202...,92e82ea2-e1f3-4c2d-a284-151efd99281e,5df36281-6dc6-4d5d-89a7-5fcfa86f7608,/items/4521724,2024-10-05 22:14:10.089116+00,"Səbail Rayonu, İzzət Nəbiyev küçəsi, Fəxri Xiy...",3 450 AZN/m²,499999.0,AZN,Kamran,mülkiyyətçi,NaN,NaN,İzzət Nəbiyev küç.,40.358817,49.824092,yeniləndi: dünən 23:52,1155,05.10.2024,23:52,92e82ea2-e1f3-4c2d-a284-151efd99281e,NaN,Köhnə tikili,7 / 9,4.0,145 m²,NaN,var,var,NaN,92e82ea2-e1f3-4c2d-a284-151efd99281e,/items/4521724,Şəhidlər xiyabanı * Dağüstü parkı * Səbail r.
1,883e20f0-8872-49a5-8b4b-63b8301b5f8f,/alqi-satqi?page=250,/items/4669294,2024-10-05 22:07:37.60613+00,77000.0,AZN,Biləcəri q.,"4 otaqlı, 90 m²","Bakı, dünən 23:56",bakı,05.10.2024,23:56,Təmirli,NaN,NaN,NaN,NaN,NaN,https://bina.azstatic.com/uploads/f460x345/202...,505eaf81-6bc8-4094-9b00-aa82066548ee,883e20f0-8872-49a5-8b4b-63b8301b5f8f,/items/4669294,2024-10-05 22:14:10.089116+00,"Biləcəridə Abidəyə yaxin 91,92,202 saylı marşr...",NaN,77000.0,AZN,Dasinmaz Emlak,vasitəçi (agent),NaN,NaN,Biləcəri qəs.,40.420897,49.807035,yeniləndi: 04 oktyabr 2024,218,04.10.2024,NaN,505eaf81-6bc8-4094-9b00-aa82066548ee,NaN,Həyət evi/Bağ evi,NaN,4.0,90 m²,1.3 sot,var,yoxdur,NaN,505eaf81-6bc8-4094-9b00-aa82066548ee,/items/4669294,Binəqədi r.* Biləcəri q.
2,55c36fb1-a3af-476e-ba17-a81f6795be8d,/alqi-satqi?page=250,/items/4669293,2024-10-05 22:07:37.60613+00,92000.0,AZN,İnşaatçılar m.,"3 otaqlı, 60 m²","Bakı, dünən 23:55",bakı,05.10.2024,23:55,Təmirli,NaN,NaN,NaN,Çıxarış var,NaN,https://bina.azstatic.com/uploads/f460x345/202...,fa63b201-999d-43b5-a61a-778d9d79a6c6,55c36fb1-a3af-476e-ba17-a81f6795be8d,/items/4669293,2024-10-05 22:14:10.089116+00,Salam əleykum. \nİnşaatçılar metrosuna yaxın m...,NaN,92000.0,AZN,Məhəmməd,vasitəçi (agent),NaN,NaN,Mirzə Cabbar Məmmədzadə küç.,40.390293,49.802656,yeniləndi: 04 oktyabr 2024,190,04.10.2024,NaN,fa63b201-999d-43b5-a61a-778d9d79a6c6,NaN,Həyət evi/Bağ evi,NaN,3.0,60 m²,0.1 sot,var,var,NaN,fa63b201-999d-43b5-a61a-778d9d79a6c6,/items/4669293,İnşaatçılar m.* Yasamal r.
3,acf1aa8d-a46a-40f5-b6f2-f7a449569337,/alqi-satqi?page=250,/items/4647811,2024-10-05 22:07:37.60613+00,95000.0,AZN,Qaraçuxur q.,130 m²,"Bakı, dünən 23:55",bakı,05.10.2024,23:55,Təmirli,vipped,featured,NaN,Çıxarış var,İpoteka var,https://bina.azstatic.com/uploads/f460x345/202...,5db56980-05cc-4925-b55b-f58fb3f4d2b6,acf1aa8d-a46a-40f5-b6f2-f7a449569337,/items/4647811,2024-10-05 22:14:10.089116+00,Barter maraqlidir üstünlük maşina verilir\nHər...,NaN,95000.0,AZN,Elçin,mülkiyyətçi,NaN,NaN,Qaraçuxur qəs.,40.393614,49.981553,yeniləndi: 04 oktyabr 2024,1314,04.10.2024,NaN,5db56980-05cc-4925-b55b-f58fb3f4d2b6,NaN,Obyekt,NaN,NaN,130 m²,NaN,var,var,var,5db56980-05cc-4925-b55b-f58fb3f4d2b6,/items/4647811,Suraxanı r.* Qaraçuxur q.
4,22d840df-9283-4112-bc71-7432511fc776,/alqi-satqi?page=250,/items/4638863,2024-10-05 22:07:37.60613+00,220000.0,AZN,Əhmədli m.,"3 otaqlı, 100 m², 15/16 mərtəbə","Bakı, dünən 23:52",bakı,05.10.2024,23:52,Təmirli,NaN,NaN,Agentlik,Çıxarış var,NaN,https://bina.azstatic.com/uploads/f460x345/202...,d71cf9a9-86dd-4162-b540-6252a8659a09,22d840df-9283-4112-bc71-7432511fc776,/items/4638863,2024-10-05 22:14:10.089116+00,Əhmədli qəs. Qaçaq Nəbi küçəsi 3 otaga duzelm...

In [77]:
df_clean = df.copy()
df_clean = df_clean.drop(columns=["Binanın növü", "hour_y"])

In [78]:
df_clean["bill_of_sale_flag"] = df_clean["Çıxarış"].str.strip().str.lower().map({"var": 1, "yoxdur": 0})
df_clean = df_clean.drop(columns = ["featured", "vip", "mortgage", "İpoteka", "bill_of_sale", "Çıxarış", "repair"])
df_clean["Təmir"] = df_clean["Təmir"].fillna("Unknown")

In [79]:
def seller_type(row):
    if pd.notnull(row["shop_name"]):
        return "Agency"
    elif pd.notnull(row["owner_name"]):
        return "Individual"
    return "Unknown"

In [80]:
df_clean["seller_type"] = df_clean.apply(seller_type, axis=1)

In [81]:
def parse_area(val):
    if pd.isnull(val):
        return np.nan
    val = val.strip()
    if "sot" in val:
        return float(val.replace("sot", "").strip().replace(",", ".")) * 100
    return float(val.replace("m²", "").strip().replace(" ", ""))


In [82]:
df_clean["Sahə_numeric"] = df_clean["Sahə"].apply(parse_area)

In [83]:
applicable_categories = ["Köhnə tikili", "Yeni tikili", "Ofis", "Həyət evi/Bağ evi"]
mask = df_clean["Kateqoriya"].isin(applicable_categories)
medians = df_clean.loc[mask].groupby("Kateqoriya")["Otaq sayı"].median()

In [84]:
for cat, med in medians.items():
    idx = df_clean[(df_clean["Kateqoriya"] == cat) & (df_clean["Otaq sayı"].isnull())].index
    df_clean.loc[idx, "Otaq sayı"] = med

In [85]:
df_clean = df_clean[df_clean["price"] >= 1000].copy()

In [86]:
def iqr_bounds(series, k=1.5):
    q1, q3 = series.quantile([0.25, 0.75])
    iqr = q3 - q1
    return q1 - k * iqr, q3 + k * iqr

In [87]:
def cap_grouped(df, col, group_col="Kateqoriya", k=1.5):
    capped = df[col].copy()
    for cat, group in df.groupby(group_col):
        low, high = iqr_bounds(group[col].dropna(), k)
        idx = group.index
        capped.loc[idx[df.loc[idx, col] < low]] = low
        capped.loc[idx[df.loc[idx, col] > high]] = high
    return capped

In [88]:
df_clean["price_capped"] = cap_grouped(df_clean, "price")
df_clean["Sahə_numeric_capped"] = cap_grouped(df_clean, "Sahə_numeric")

In [89]:
low_v, high_v = iqr_bounds(df_clean["views"])
df_clean["views_capped"] = df_clean["views"].clip(low_v, high_v)

In [90]:
floor_split = df_clean["Mərtəbə"].str.split("/", expand=True)
df_clean["floor_current"] = pd.to_numeric(floor_split[0].str.strip(), errors="coerce")
df_clean["floor_total"] = pd.to_numeric(floor_split[1].str.strip(), errors="coerce")
df_clean["floor_ratio"] = (df_clean["floor_current"] / df_clean["floor_total"]).round(2)

In [91]:
def loc_type(loc):
    if pd.isnull(loc):
        return "Unknown"
    match = re.search(r"\s(r|m|q)\.$", loc.strip())
    mapping = {"r": "District center", "m": "Near metro", "q": "Settlement"}
    return mapping.get(match.group(1), "Other") if match else "Other"

In [92]:
df_clean["location_type"] = df_clean["location"].apply(loc_type)

In [93]:
feature_columns = ["Sahə_numeric_capped", "Otaq sayı", "views_capped", "floor_ratio",
                    "Kateqoriya", "location_type", "seller_type", "Təmir", "city"]
target_column = "price_capped"

In [94]:
df_model = df_clean[feature_columns + [target_column]].copy()
X = df_model[feature_columns]
y = df_model[target_column]

In [95]:
# Same split as every previous checkpoint - same random_state, so this is the same X_test
# that has never been touched until now.
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 42)

In [96]:
numeric_features = ["Sahə_numeric_capped", "Otaq sayı", "views_capped", "floor_ratio"]
categorical_features = ["Kateqoriya", "location_type", "seller_type", "Təmir", "city"]

In [97]:
numeric_pipeline = Pipeline(steps = [
    ("impute", SimpleImputer(strategy = "median")),
    ("scale", StandardScaler()),
])
categorical_pipeline = Pipeline(steps = [
    ("impute", SimpleImputer(strategy = "constant", fill_value = "Unknown")),
    ("encode", OneHotEncoder(handle_unknown = "ignore")),
])
preprocessor = ColumnTransformer(transformers = [
    ("num", numeric_pipeline, numeric_features),
    ("cat", categorical_pipeline, categorical_features),
])

In [98]:
# Best hyperparameters found by RandomizedSearchCV in Checkpoint 4
final_model = Pipeline(steps = [
    ("preprocessor", preprocessor),
    ("model", RandomForestRegressor(
        n_estimators = 150,
        max_depth = 20,
        min_samples_leaf = 1,
        max_features = 0.5,
        random_state = 42,
        n_jobs = -1,
    )),
])

In [99]:
print("X_train:", X_train.shape, "| X_test:", X_test.shape)

X_train: (80612, 9) | X_test: (20153, 9)


## 2. Fit the final model on the full training set

This uses the tuned hyperparameters from Checkpoint 4, fit once on all of X_train (not just a CV fold), this is the actual final model.


In [100]:
final_model.fit(X_train, y_train)
print("Final model fitted on the full training set.")


Final model fitted on the full training set.


## 3. Evaluate on the test set


In [101]:
test_predictions = final_model.predict(X_test)

In [102]:
test_mae = mean_absolute_error(y_test, test_predictions)
test_rmse = np.sqrt(mean_squared_error(y_test, test_predictions))
test_r2 = r2_score(y_test, test_predictions)
test_mape = (np.abs((y_test - test_predictions) / y_test)).mean() * 100

In [103]:
print("Final test set performance:")
print(f"  MAE : {test_mae:,.0f} AZN")
print(f"  RMSE: {test_rmse:,.0f} AZN")
print(f"  R2  : {test_r2:.3f}")
print(f"  MAPE: {test_mape:.1f}%")

Final test set performance:
  MAE : 56,831 AZN
  RMSE: 133,719 AZN
  R2  : 0.784
  MAPE: 39.1%


In [104]:
price_bucket = pd.qcut(y_test, 5)
ape = np.abs((y_test - test_predictions) / y_test) * 100
mape_by_bucket = pd.Series(ape.values, index = y_test.index).groupby(price_bucket, observed = True).mean().round(1)

In [105]:
print("MAPE by price quintile:")
print(mape_by_bucket)

MAPE by price quintile:
price_capped
(999.999, 132000.0]      123.6
(132000.0, 185000.0]      20.2
(185000.0, 255000.0]      17.6
(255000.0, 368000.0]      15.6
(368000.0, 3000000.0]     17.9
dtype: float64


In [106]:
print(f"Overall mean MAPE: {ape.mean():.1f}%")
print(f"Median APE: {np.median(ape):.1f}%")

Overall mean MAPE: 39.1%
Median APE: 12.1%


## 4. Compare against the Checkpoint 4 cross-validation estimate

In [107]:
cv_estimate = {"MAE": 57159, "RMSE": 134477, "R2": 0.773}  # from Checkpoint 4, tuned model, 5-fold CV mean

In [108]:
comparison_df = pd.DataFrame([
    {"stage": "CV estimate (Checkpoint 4, train only)", "MAE": f"{cv_estimate['MAE']:,.0f}",
     "RMSE": f"{cv_estimate['RMSE']:,.0f}", "R2": f"{cv_estimate['R2']:.3f}"},
    {"stage": "Test set (this checkpoint, seen once)", "MAE": f"{test_mae:,.0f}",
     "RMSE": f"{test_rmse:,.0f}", "R2": f"{test_r2:.3f}"},
])

In [109]:
comparison_df

,stage,MAE,RMSE,R2
0,"CV estimate (Checkpoint 4, train only)","57,159","134,477",0.773
1,"Test set (this checkpoint, seen once)","56,831","133,719",0.784


the test-set MAE (56,831 AZN) and R2 (0.784) land almost exactly on the Checkpoint 4 cross-validation estimate (57,159 AZN / 0.773), only a 0.6% gap on MAE. This is the confirmation that the CV process wasn't misleading. There's no sign of the test-set-contamination problem this project was specifically checking for, and the model should perform similarly on new, real-world listings it wasn't trained on.

MAPE came out at 39.1%, well above the "within about 10-15%" aspiration written in Checkpoint 1. That earlier number was a rough directional guess before any model existed, not a guarantee, and the actual result shows the gap plainly rather than hiding it. A likely explanation is the very wide price range in this dataset. MAPE is an average tail-sensitive percentage, so a model can have a strong R2 and still carry a high average percentage error on the cheaper listings, where even a moderate AZN error is a large percentage of the price. This is worth stating clearly in the business report rather than only reporting the more flattering R2/MAE figures.